### PR TITLE
OpenAI supports embeddings, and storage supports streams and overwrite

### PR DESCRIPTION
**Pull request by christothes**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAiModelDeployment.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+public class OpenAiModelDeployment
```
Comment (KrzysztofCwalina) [inline sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAiModelDeployment.cs:4]: Could we just call it AIModel? The current name is mouthful

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs [hunk 2]
```diff
@@ -8,21 +8,30 @@
...
 
 public class OpenAIFeature : CloudMachineFeature
 {
-    public string Model { get; }
-    public string ModelVersion { get; }
+    private OpenAiModelDeployment _chatDeployment;
+    private OpenAiModelDeployment? _embeddingsDeployment;
 
-    public OpenAIFeature(string model, string modelVersion) {  Model = model; ModelVersion = modelVersion; }
+    public OpenAIFeature(OpenAiModelDeployment chatDeployment, OpenAiModelDeployment? embeddingsDeployment = default)
```
Comment (KrzysztofCwalina) [inline sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs:20]: ultimately we need to support a collection of deployments, and so I am not thrilled with special casing for 2 deployments.

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs [hunk 3]
```diff
@@ -8,21 +8,30 @@
...
+        _chatDeployment = chatDeployment;
+        _embeddingsDeployment = embeddingsDeployment;
+    }
 
     public override void AddTo(CloudMachineInfrastructure cloudMachine)
     {
         CognitiveServicesAccount cognitiveServices = new("openai")
         {
-            Name = cloudMachine.Id,
+            Name = $"{cloudMachine.Id}/{cloudMachine.Id}",
```
Comment (KrzysztofCwalina) [inline sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs:34]: why this change?

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAiModelDeployment.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+{
+    public OpenAiModelDeployment(string model, string modelVersion) { Model = model; ModelVersion = modelVersion; }
+    public string Model { get; }
+    public string ModelVersion { get; }
```
Comment (KrzysztofCwalina) [inline sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAiModelDeployment.cs:8]: maybe we also need a property to specify if the model is for chat, embeddings, etc.

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/OFX/StorageServices.cs [hunk 5]
```diff
@@ -42,14 +42,50 @@ private BlobContainerClient GetContainer(string containerName)
         return container;
     }
 
-    public string UploadJson(object json, string? name = default)
+    public string UploadJson(object json, string? name = default, bool overwrite = false)
```
Comment (KrzysztofCwalina) [inline sdk/provisioning/Azure.Provisioning.CloudMachine/src/OFX/StorageServices.cs:45]: could you also set content type in the blob metadata (same with upload bytes).

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs (hunk 2 above)
Comment (christothes) [reply sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs:20]: I started down that path, but there are some problems to be solved with how the models are represented in `GetConnectionOptions`. This approach allows us to retain some reasonable defaults. I think we can refactor this as the need arises for more deployed model types.

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs (hunk 3 above)
Comment (christothes) [reply sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs:34]: I noticed that this is how AI Studio names them when it deploys the models.

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/OFX/StorageServices.cs (hunk 5 above)
Comment (christothes) [reply sdk/provisioning/Azure.Provisioning.CloudMachine/src/OFX/StorageServices.cs:45]: Will do, but we don't know the content type with upload bytes.

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply sdk/provisioning/Azure.Provisioning.CloudMachine/src/AzureSdkExtensions/OpenAIFeature.cs:20]: I think it's fine to do this as a separate PR.

File: sdk/provisioning/Azure.Provisioning.CloudMachine/src/OFX/StorageServices.cs (hunk 5 above)
Comment (KrzysztofCwalina) [reply sdk/provisioning/Azure.Provisioning.CloudMachine/src/OFX/StorageServices.cs:45]: bytes should be octet stream by default. If the user wants something else, they should pass BinaryData with ContentType set

**Review by KrzysztofCwalina — APPROVED** [review]
